### PR TITLE
Populate column selector for datasets which are not ready yet

### LINKED ISSUE #3758: Dynamically Generated Tool Parameters cannot be set until dataset has finished()
You used to be able to manually specify values when the input dataset was not yet ready.

Example:
 - Upload Tabular File, explicitly set datatype to `tabular`.
 - Try to queue up `Group` tool.
 - Cannot queue up tool, because you cannot specify column parameters.

The user was previously presented with a text box to allow manual entry.

### PR DESCRIPTION
**Pull request by guerler**
Resolve #3758 for column parameters. ping @blankenberg.

### CODE REVIEW

File: lib/galaxy/tools/parameters/basic.py [hunk 1]
```diff
@@ -1114,11 +1116,16 @@ def get_column_list(self, trans, other_values):
...
+            if not hasattr(dataset, 'metadata') or not hasattr(dataset.metadata, 'columns'):
                 return []
             # Build up possible columns for this dataset
             this_column_list = []
-            if self.numerical:
+            # Valid column-based datasets contain at least 1 column if that column has not been
+            # specified we prepopulate the selector assuming that the datasets is not ready yet.
+            if dataset.metadata.columns is None:
+                for i in range(0, MAX_DEFAULT_COLUMNS):
+                    this_column_list.append(str(i + 1))
```
Comment (nsoranzo) [inline lib/galaxy/tools/parameters/basic.py]: You can replace the last 2 lines with
```python
                this_column_list = list(map(str, range(1, MAX_DEFAULT_COLUMNS + 1)))
```
which takes around half the time.
Similarly for lines 1134-1135.

Comment (guerler) [reply lib/galaxy/tools/parameters/basic.py]: Thx!

**Review by nsoranzo — APPROVED** [review]

**Comment by blankenberg** [comment]
Awesome! Thanks. Will check it out soon.

**Comment by blankenberg** [comment]
Couple of concerns here. 

I have a tool that takes a tabular input file and creates a tabular output file where the initial metadata of the output file is populated with a copy from the input dataset (`metadata_source`; this is ok). When when the job finishes, the output columns are correctly set to their real values. But I currently can still not select the newly created column because this change doesn't consider dataset set, only metadata presence.

Hardcoding 999 columns seems like a bandaid at best. This does not allow selecting e.g. column 1005 from the still running dataset.

I think we need to allow users to manually enter values (like it used to) and then validate them before the tool run.

See the `Add a column` tool, followed by trying to select the new column with e.g. the `join by specified field` tool.

**Comment by guerler** [comment]
Alright, Ill be revisiting this issue.

**Comment by blankenberg** [comment]
I agree it is nice to have the dropdown, since in many cases the number of columns might not change, but I wonder if there is a smooth way to allow both?

**Pull request merged**